### PR TITLE
distsql: fix span leak in outbox

### PR DESCRIPTION
**Pull request by asubiotto**
Release note (bug fix): Fix a small memory leak when running distributed
queries.
cc @nvanbenschoten

### CODE REVIEW

File: pkg/sql/distsqlrun/outbox.go [hunk 1]
```diff
@@ -283,6 +283,9 @@ func (m *outbox) mainLoop(ctx context.Context) error {
 							return err
 						}
 					}
+				} else {
+					// Case in which we are not recording but still need to finish span.
+					tracing.FinishSpan(span)
```
Comment (jordanlewis) [inline pkg/sql/distsqlrun/outbox.go]: Can this be done in a `defer` instead, or is it not possible?

Comment (RaduBerinde) [reply pkg/sql/distsqlrun/outbox.go]: We want to Finish it before calling `getTraceData` above.. But I think we could pull out the `FinishSpan` and `getTraceData` code out of the `if !ok {}`. If we're not recording, `getTraceData` would just return `nil`.

Comment (RaduBerinde) [reply pkg/sql/distsqlrun/outbox.go]: Also, are we leaking the span in the `return nil` case above (line 273)?

Comment (asubiotto) [reply pkg/sql/distsqlrun/outbox.go]: I think pulling it out (as well as the flush) would work. Good catch on line 273

Comment (asubiotto) [reply pkg/sql/distsqlrun/outbox.go]: Done. Logging errors now (should they be info?).

File: pkg/sql/distsqlrun/outbox.go [hunk 2]
```diff
@@ -203,6 +203,18 @@ func (m *outbox) mainLoop(ctx context.Context) error {
...
 	}
+	defer func() {
+		tracing.FinishSpan(span)
+		if trace := getTraceData(ctx); trace != nil {
+			err := m.addRow(ctx, nil, &ProducerMetadata{TraceData: trace})
+			if err != nil {
+				log.Error(ctx, err)
+			}
+		}
+		if err := m.flush(ctx); err != nil {
```
Comment (RaduBerinde) [inline pkg/sql/distsqlrun/outbox.go]: Are we sure this (and `addRow`) is ok to do in all the `return` cases in the function?

File: pkg/sql/distsqlrun/outbox.go [hunk 3]
```diff
@@ -275,16 +287,10 @@ func (m *outbox) mainLoop(ctx context.Context) error {
...
-					tracing.FinishSpan(span)
-					if trace := getTraceData(ctx); trace != nil {
-						err := m.addRow(ctx, nil, &ProducerMetadata{TraceData: trace})
-						if err != nil {
-							return err
-						}
-					}
 				}
-				return m.flush(ctx)
+				return nil
```
Comment (RaduBerinde) [inline pkg/sql/distsqlrun/outbox.go]: We used to return the flush error in this case..

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31736)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsqlrun/outbox.go, line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31736#-LPWttUW7a-PzDrPGWyk-r1-214:-LPWu0a39oOztzXusr6Y:b-t1f3lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/661ad5116f99b25b85f808e28019dd886ecd90bd/pkg/sql/distsqlrun/outbox.go#L214)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Are we sure this (and `addRow`) is ok to do in all the `return` cases in the function?
</blockquote></details>

Good point. I went with a `spanFinished` boolean instead and kept the `addRow` and `flush` in the original code path.
___
*[pkg/sql/distsqlrun/outbox.go, line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31736#-LPWttUW7a-PzDrPGWyk-r1-293:-LPX2AYh0e9U7HHh0C6Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/661ad5116f99b25b85f808e28019dd886ecd90bd/pkg/sql/distsqlrun/outbox.go#L293)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We used to return the flush error in this case..
</blockquote></details>

Done.

**Comment by RaduBerinde** [comment]
Oh wow this version is so much cleaner! LGTM!

**Comment by nvanbenschoten** [comment]
LGTM, thanks for this quick fix @asubiotto!

**Pull request merged**